funnel/models/post.py

```diff
@@ -0,0 +1,265 @@
+        return '<Post "{title}" {uuid_b58}'.format(
+            title=self.title, uuid_b58=self.uuid_b58
+        )
+
+    state.add_conditional_state(
+        'UNPUBLISHED',
+        state.DRAFT,
+        lambda post: post.published_at is not None,
+        lambda post: post.published_at.isnot(None),
+        label=('unpublished', __("Unpublished")),
+    )
+
+    @with_roles(call={'editor'})
+    @state.transition(
+        state.DRAFT,
+        state.PUBLISHED,
+        title=__("Publish post"),
+        message=__("Post has been published"),
+    )
+    def publish(self, actor):
+        self.published_by = actor
+        if self.published_at is None:
+            self.published_at = db.func.utcnow()
+
+    @with_roles(call={'editor'})
+    @state.transition(
+        state.PUBLISHED,
+        state.DRAFT,
+        title=__("Undo publish"),
+        message=__("Post is now a draft"),
+    )
+    def undo_publish(self):
+        pass
+
+    @with_roles(call={'creator', 'editor'})
+    @state.transition(
+        None,
+        state.DELETED,
+        title=__("Delete post"),
+        message=__("Post has been deleted"),
+    )
+    def delete(self, actor):
+        if self.state.DRAFT:
+            # If it's a draft post, hard delete it
+            db.session.delete(self)
+        else:
+            # If not, then soft delete
+            self.deleted_by = actor
+            self.deleted_at = db.func.utcnow()
+
+    @with_roles(call={'editor'})
+    @state.transition(
+        state.DELETED,
+        state.DRAFT,
+        title=__("Undo delete"),
+        message=__("Post is now a draft"),
+    )
+    def undo_delete(self):
+        self.deleted_by = None
+        self.deleted_at = None
+
+    @with_roles(call={'editor'})
+    @visibility_state.transition(
+        visibility_state.RESTRICTED,
+        visibility_state.PUBLIC,
+        title=__("Make post public"),
+        message=__("Post is now public"),
+    )
+    def make_public(self):
+        pass
+
+    @with_roles(call={'editor'})
+    @visibility_state.transition(
+        visibility_state.PUBLIC,
+        visibility_state.RESTRICTED,
+        title=__("Make post restricted"),
+        message=__("Post is now restricted"),
+    )
+    def make_restricted(self):
+        pass
+
+    def roles_for(self, actor=None, anchors=()):
+        roles = super().roles_for(actor, anchors)
+        project_roles = (
+            self.project.roles_for(actor) if self.project is not None else set()
+        )
+        profile_roles = (
+            self.profile.roles_for(actor) if self.profile is not None else set()
+        )
+
+        if self.visibility_state.RESTRICTED:
+            if 'participant' in project_roles or 'admin' in profile_roles:
+                roles.add('reader')
+        else:
+            roles.add('reader')
+
+        if 'editor' in project_roles:
+            roles.add('editor')
+        if 'admin' in profile_roles:
+            roles.add('admin')
```

Comment:
All this can be in `grants_via`. Using the declarative approach will aid performance since we won't check for unnecessary roles in each request.

Reply:
Forgot to remove this.